sc-memory/sc-memory/src/scs/scs_parser.cpp

```diff
@@ -413,44 +456,125 @@ ElementHandle Parser::ProcessIdentifier(std::string const & name)
-    if (connectorEl.GetType() == ScType::ConstPermPosArc && scs::TypeResolver::IsKeynodeType(idtf))
-    {
-      ParsedElement & targetEl = GetParsedElementRef(trg);
-      ScType const newType = targetEl.m_type | scs::TypeResolver::GetKeynodeType(idtf);
+  auto AddConnector =
+      [this, &connector](
+          ElementHandle const & sourceHdl, ElementHandle const & connectorHdl, ElementHandle const & targetHdl)
+  {
+    ParsedElement const & source = GetParsedElement(sourceHdl);
+    ParsedElement & target = GetParsedElementRef(targetHdl);
 
-      if (targetEl.m_type.CanExtendTo(newType))
-        targetEl.m_type = newType;
+    if (source.IsElementType() && (connector.GetType() == ScType::ConstPermPosArc))
+    {
+      std::string const & sourceIdtf = source.GetIdtf();
+      ScType const sourceType = scs::TypeResolver::GetElementType(sourceIdtf);
+      ScType const targetType = target.m_type;
+      ScType const newTargetType = targetType | sourceType;
+
+      if (target.IsElementType() && !newTargetType.CanExtendTo(ScType::ConstNodeClass))
+        SC_THROW_EXCEPTION(
+            utils::ExceptionParseError,
+            "Can't extend type `" << std::string(targetType) << "` using type `" << std::string(sourceType)
+                                  << "` for specified sc-element, because sc-element `" << target.GetIdtf()
+                                  << "` is sc-element denoting type of sc-elements.");
       else
-        SC_THROW_EXCEPTION(utils::ExceptionParseError, "Can't merge types for element " + targetEl.GetIdtf());
+      {
+        // Check that any type of target sc-element without constancy mask can be extended to type of sc-elements 
+        // represented by source sc-element.
+        ScType const & targetTypeWithoutConstancyMask = targetType.BitAnd(~(ScType::Const | ScType::Var));
+        bool canTargetTypeWithoutConstancyMaskExtendToSourceType 
+            = targetTypeWithoutConstancyMask.CanExtendTo(sourceType);
+
+        // Check that any type of source sc-element with constancy mask can be extended to type of sc-elements 
+        // represented by source sc-element with constancy mask.
+        ScType const & sourceTypeWithConstancyMask = sourceType.BitAnd(ScType::Const | ScType::Var);
+        ScType const & targetTypeWithConstancyMask = targetType.BitAnd(ScType::Const | ScType::Var);
+        bool canSourceTypeWithConstancyMaskExtendToTargetTypeWithConstancyMask =
+            sourceTypeWithConstancyMask.CanExtendTo(targetTypeWithConstancyMask);
+
+        if (canTargetTypeWithoutConstancyMaskExtendToSourceType
+            && canSourceTypeWithConstancyMaskExtendToTargetTypeWithConstancyMask)
+          target.m_type = newTargetType;
+        else if (!sourceType.CanExtendTo(targetType))
+          SC_THROW_EXCEPTION(
+              utils::ExceptionParseError,
+              "Can't extend type `" << std::string(targetType) << "` using type `" << std::string(sourceType)
```

Comment:
I got
Can't extend type `MembershipArc` using type `ConstPermPosArc` for specified sc-element.
when I tried to build

@arc = (... ?.?> ...);;
sc_main_arc -> @arc;;